pgjdbc/src/main/java/org/postgresql/sasl/ScramAuthenticator.java

```diff
@@ -0,0 +1,154 @@
+    this.user = user;
+    this.password = password;
+    this.pgStream = pgStream;
+  }
+
+  public void processServerMechanismsAndInit() throws IOException, PSQLException {
+    String scramAuthenticationMechanism = pgStream.receiveString();
+    pgStream.receiveChar(); // '0' terminating the String
+    if (!EXPECTED_SERVER_ADVERTISED_METHODS.equals(scramAuthenticationMechanism)) {
+      throwConnectionRejectedException(
+          "Unsupported SCRAM mechanism(s) '{0}' advertised by the server",
+          scramAuthenticationMechanism
+      );
+    }
+    logFinest(" Using SCRAM mechanism {0}", EXPECTED_SERVER_ADVERTISED_METHODS);
+
+    scramSession =
+        scramClient.scramSession("*");   // Real username is ignored by server, uses startup one
+  }
+
+  public void sendScramClientFirstMessage() throws IOException {
+    String clientFirstMessage = scramSession.clientFirstMessage();
+    logFinest(" FE=> SASLInitialResponse( {0} )", clientFirstMessage);
+
+    String scramMechanismName = scramClient.getScramMechanism().getName();
+    authenticationMessageSender(
+        s -> {
+          s.send(scramMechanismName.getBytes(StandardCharsets.UTF_8));
+          s.sendChar(0); // List terminated in '\0'
+          s.sendInteger4(clientFirstMessage.length());
+          s.send(clientFirstMessage.getBytes(StandardCharsets.UTF_8));
+        },
+        (scramMechanismName.length() + 1) + Integer.BYTES + clientFirstMessage.length()
+    );
+  }
+
+  public void processServerFirstMessage(int length) throws IOException, PSQLException {
+    String serverFirstMessage = pgStream.receiveString(length);
+    logFinest(" <=BE AuthenticationSASLContinue( {0} )", serverFirstMessage);
+
+    try {
+      serverFirstProcessor = scramSession.receiveServerFirstMessage(serverFirstMessage);
+    } catch (ScramException e) {
+      throwConnectionRejectedException("Invalid server-first-message: {0}", serverFirstMessage);
+    }
+    logFinest(
+        " <=BE AuthenticationSASLContinue(salt={0}, iterations={1})",
+        serverFirstProcessor.getSalt(), serverFirstProcessor.getIteration()
+    );
+
+    clientFinalProcessor = serverFirstProcessor.clientFinalProcessor(password);
+
+    String clientFinalMessage = clientFinalProcessor.clientFinalMessage();
+    logFinest(" FE=> SASLResponse( {0} )", clientFinalMessage);
+
+    authenticationMessageSender(
+        s -> s.send(clientFinalMessage.getBytes(StandardCharsets.UTF_8)),
+        clientFinalMessage.length()
+    );
+  }
+
+  public void verifyServerSignature(int length) throws IOException, PSQLException {
+    String serverFinalMessage = pgStream.receiveString(length);
+    logFinest(" <=BE AuthenticationSASLFinal( {0} )", serverFinalMessage);
+
+    try {
+      clientFinalProcessor.receiveServerFinalMessage(serverFinalMessage);
+    } catch (ScramParseException e) {
+      throwConnectionRejectedException("Invalid server-final-message: {0}", serverFinalMessage);
```

Comment:
Should the original exception be tied to the thrown one as a `cause`? The same for two other `catch` blocks below.